### ISSUE TITLE
TypeError: Cannot read property '_walk' of null

### ISSUE THREAD

**Issue by joshsearles** [terser-js/terser#120]
Bug Report

terser v3.9.2

Building Ember js Project

```
Error: Build Canceled: Broccoli Builder ran into an error with `UglifyWriter` plugin. 💥
Cannot read property '_walk' of null
TypeError: Cannot read property '_walk' of null
    at AST_Node.eval (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1145:28)
    at Object.TreeWalker._visit (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1546:21)
    at AST_Node.DEFNODE._walk (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1144:24)
    at AST_Node.eval (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1091:40)
    at Object.TreeWalker._visit (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1546:21)
    at AST_Node.DEFNODE._walk (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1089:24)
    at AST_Node.eval (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:999:32)
    at Object.TreeWalker._visit (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1546:21)
    at AST_Node.DEFNODE._walk (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:996:24)
    at walk_body (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:510:17)Pipeline aborted
```

EDIT:
to replicate on a basic ember js project (apparently it was the ember-cli-deploy plugin for me)
```
# npm install ember-cli-deploy
# npm install && ember deploy production
```

**Comment by fabiosantoscode** [comment]
Try to generate a minimal code example which fails on terser

**Comment by alechirsch** [comment]
+1, I also run into this problem, these other issues are related

https://github.com/angular/angular-cli/issues/12414
https://github.com/ember-cli/ember-cli/issues/8075

**Comment by kzc** [comment]
@alechirsch Without a standalone javascript example reproducing the problem nothing can be done. All the information outlined here is required:

https://raw.githubusercontent.com/terser-js/terser/master/.github/ISSUE_TEMPLATE.md

**Comment by jrjohnson** [comment]
I'm attempting to track down where this issue may fall. It did not exist in `3.8.2` and I'm seeing it in `3.9.2`. Unfortunately all of these releases are on NPM, but tags have no been pushed to github. If someone can push those tags it will be much easier to investigate this issue.

**Comment by kzc** [comment]
To make a bug repro just provide the arguments passed to `terser` `minify()`.

The easiest way to do that is to find the `terser` directory in `node_modules`:
```
find . -name terser -type d
./node_modules/terser
```
then go to that directory and apply this patch:

```patch
--- a/lib/minify.js
+++ b/lib/minify.js
@@ -59,2 +59,4 @@ function to_json(cache) {
 function minify(files, options) {
+    console.error("*** terser options:\n", options);
+    console.error("*** terser input:\n", files);
     var warn_function = AST_Node.warn_function;
```
and then run your project's build command as normal.

Paste the stderr output into a gist and drop the URL here in this ticket.

**Comment by jrjohnson** [comment]
Thanks, I've pasted the (huge) output into https://gist.github.com/jrjohnson/0a73a3f8588b5ced74f79530eda2a449

At the end it refers to a Stack Trace and Error Report and I put that at https://gist.github.com/jrjohnson/cf4cbcb9922d5f9a1286b41f1eaee079

**Comment by kzc** [comment]
@jrjohnson Unable to reproduce any error using `terser@3.9.2` with the minify options:
```
 { mangle: { safari10: true },
  compress: { negate_iife: false, sequences: 30 },
  output: { semicolons: false },
  sourceMap: false }
```
and the (cleaned up) source from https://gist.github.com/jrjohnson/0a73a3f8588b5ced74f79530eda2a449

Strangely, the gist link above mentions an "Invalid assignment" parse error, which is different than the "Cannot read property '_walk' of null" error reported in the top post. I was not able to reproduce either error.

**Comment by jrjohnson** [comment]
I'm sorry @kzc I think I may have been in a rush and gotten sloppy with my original output. I've replaced both gists with an update that is failing with the Cannot read property '_walk' of null error.

**Comment by kzc** [comment]
It's possible that code is being minified in parallel by ember. If so, disable this option.

**Comment by kzc** [comment]
@jrjohnson That did the trick.

Here's a reduced test case:

```
$ cat bug.js 
function foo(node) {
    var traverse = function(obj) {
        var i = obj.data;
        return i && i.a != i.b;
    };
    while (traverse(node)) {
        node = node.data;
    }
    return node;
}
var x = {
    a: 1,
    b: 2,
    data: {
        a: "hello",
    }
};
console.log(foo(x).a, {a : "world"}.a);
```
Expected:
```
$ cat bug.js | node
hello world

$ cat bug.js | terser | node
hello world
```
Actual:
```
$ cat bug.js | terser -c
ERROR: Cannot read property '_walk' of null
```
This is not a recent bug. It is present in both `uglify-es` and latest `uglify-js` introduced in 9b58b54e2d7ce4ab2c015611add52c7475de9eab:
```
$ uglify-es -V
uglify-es 3.3.9

$ cat bug.js | uglify-es -c
ERROR: Cannot read property '_walk' of null
```
```
$ uglify-js -V
uglify-js 3.4.9

$ cat bug.js | uglify-js -c
ERROR: Cannot read property '_walk' of null
```

**Comment by kzc** [comment]
This is an interesting bug. It is caused by the interaction of the following `compress` options:

* `collapse_vars`
* `inline=3`
* `join_vars`
* `loops`
* `reduce_vars`
* `sequences` (depending on the code)
* `unused`

Disabling any of those `compress` options will make it work. In the case of `inline`, a setting a `2` also works - which coincidentally produces the smallest Terser minified bundle for the large gist example.

What happens in the reduced test case is that the `AST_Dot.expression` incorrectly becomes `null` in `var i = obj.data;` in the course of one of the transforms which ultimately fails during a subsequent tree traversal:
```
TypeError: Cannot read property '_walk' of null
    at AST_Dot.eval 
    at TreeWalker._visit 
    at AST_Dot._walk 
    at AST_VarDef.eval 
    at TreeWalker._visit 
    at AST_VarDef._walk 
    at AST_Var.eval 
    at TreeWalker._visit 
    at AST_Var._walk 
    at walk_body 
```
I suspect that something in the `inline=3` function cloning logic went awry in relation to the loop condition.

**Comment by kzc** [comment]
@joshsearles Could you please change the title of this issue to the following?
```
TypeError: Cannot read property '_walk' of null
```

**Comment by joshsearles** [comment]
updated @kzc

**Comment by kzc** [comment]
Related `reduce_vars` / `collapse_vars` issue and its fix:

* https://github.com/mishoo/UglifyJS2/issues/1609
* https://github.com/mishoo/UglifyJS2/pull/1610

**Comment by kzc** [comment]
@jrjohnson If you're wondering why this long-standing Terser bug just appeared in the last week, by coincidence this perfectly valid September 27th [`tether@1.4.5` commit](https://github.com/HubSpot/tether/commit/4ca3d2e3cfec04549b8bc7ce426d2c14c2e393d5#diff-16aa9b816bdda8647812b6bd5bdfcffaR1256) was introduced as a dependency of your gist example which just happened to exercise it. Note its similarity to the [reduced test case](https://github.com/terser-js/terser/issues/120#issuecomment-426122195) above.

I suspect any other project using `tether@1.4.5` would have the same issue with Terser. If the Tether [function in question](https://github.com/HubSpot/tether/blob/4ca3d2e3cfec04549b8bc7ce426d2c14c2e393d5/src/js/tether.js#L768-L772) were hoisted from the `else` clause to the top of its enclosing function it'd probably avoid the Terser bug.

**Comment by jrjohnson** [comment]
I figured it must be a dependency chain issue. Either either terser's deps or in the fairly vast ember build ecosystem. Thanks for digging into it so far!  I've implemented the `collapse_vars=false` fix mentioned in some of the linked issues and I'm much happier with it as a solution now that I understand why it helps.

**Comment by fabiosantoscode** [comment]
Even more reduced test case: 

```
function foo(node) {
    var traverse = function(obj) {
        var i = obj.data;
        return i && i.a != i.b;
    };
    while (traverse(node)) { }
}   

```

**Comment by kzc** [comment]
> Even more reduced test case:

https://github.com/terser-js/terser/issues/120#issuecomment-426122195 is an `expect_stdout` test case that can be verified against a variety of [minify options](https://github.com/terser-js/terser/blob/master/test/ufuzz.json) once the bug is fixed.

**Comment by fabiosantoscode** [comment]
Even more reduced test case:

```javascript
const foo = function () {
    var t = function(o) {
        var i = o.b;
        i + i.a
    };
    while (t()) { }
}
```

**Comment by fabiosantoscode** [comment]
Fixed as of `3.10.7`

**Comment by jrjohnson** [comment]
Thank you!

**Comment by mukeshyadav** [comment]
getting same issue, using CRA2 "react": "^16.5.2",
    "react-dom": "^16.5.2",
    "react-scripts": "^2.0.5"

**Comment by saintsatplay2015** [comment]
This is an issue when trying to build Ionic 4 production apps - receiving following error message: 
`ERROR in common.00974306b5124fe008c9.js from Terser
TypeError: Cannot read property 'minify' of undefined
    at minify (/ionic-pwa/ionic-pwa/node_modules/terser-webpack-plugin/dist/minify.js:175:23)
    at module.exports (/ionic-pwa/node_modules/terser-webpack-plugin/dist/worker.js:13:40)
    at handle (/ionic-pwa/node_modules/worker-farm/lib/child/index.js:44:8)
    at process.<anonymous> (/ionic-pwa/node_modules/worker-farm/lib/child/index.js:51:3)
    at emitTwo (events.js:126:13)
    at process.emit (events.js:214:7)
    at emit (internal/child_process.js:772:12)
    at _combinedTickCallback (internal/process/next_tick.js:141:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)`

This happens when using using `Terser 3.16.0` with `terser-webpack-plugin 1.2.1`.
Have tried downgrading version of Terser to `Terser 3.14.0` with `terser-webpack-plugin 1.2.1` but this doesn't work.

How to solve this?

**Comment by jrjohnson** [comment]
The best way to get started solving this is probably to open a new issue and follow the steps in the comments above to create a reduced test case.

**Comment by filipesilva** [comment]
There is already an issue for this specific problem: https://github.com/terser-js/terser/issues/251
